In [1]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
import tensorflow as tf

%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:519: FutureWarning: Passi

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784
output_dim = nb_classes = 10
batch_size= 128
nb_epoch =20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

In [3]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

#### Model Creation with 3 hidden layers - tanh

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [4]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='tanh'))
model.add(Dense(64, activation='tanh'))
model.add(Dense(10, activation='softmax'))

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [6]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [7]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
60000/60000 [==============================] - 1s 21us/sample - loss: 1.0337 - acc: 0.7447
Epoch 2/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.5324 - acc: 0.8675
Epoch 3/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.4280 - acc: 0.8869
Epoch 4/20
60000/60000 [==============================] - 1s 20us/sample - loss: 0.3783 - acc: 0.8972
Epoch 5/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3478 - acc: 0.9042
Epoch 6/20
60000/60000 [==============================] - 1s 19us/sample - loss: 0.3262 - acc: 0.9093
Epoch 7/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3098 - acc: 0.9133
Epoch 8/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2965 - acc: 0.9165
Epoch 9/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2850 - acc: 0.9201
Epoch 10/20
60000/60000 [==============================] - 1s 19us/sample - loss: 

In [8]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.20428670691251755
Test Accuracy:0.9417999982833862


#### Model Creation with 3 hidden layers - Sigmoid

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [9]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='sigmoid'))
model.add(Dense(64, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

In [10]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_4 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_5 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [11]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [12]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.2860 - acc: 0.1847
Epoch 2/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.2229 - acc: 0.3530
Epoch 3/20
60000/60000 [==============================] - 1s 20us/sample - loss: 2.1503 - acc: 0.4455
Epoch 4/20
60000/60000 [==============================] - 1s 18us/sample - loss: 2.0475 - acc: 0.5136
Epoch 5/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.9039 - acc: 0.5617
Epoch 6/20
60000/60000 [==============================] - 1s 18us/sample - loss: 1.7234 - acc: 0.6011
Epoch 7/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.5302 - acc: 0.6514
Epoch 8/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.3504 - acc: 0.6930
Epoch 9/20
60000/60000 [==============================] - 1s 18us/sample - loss: 1.1981 - acc: 0.7284
Epoch 10/20
60000/60000 [==============================] - 1s 21us/sample - loss: 

In [13]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.5523046186447144
Test Accuracy:0.8553000092506409


#### Model Creation with 3 hidden layers - ReLU

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [14]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [15]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_8 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [16]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [17]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.2597 - acc: 0.6783
Epoch 2/20
60000/60000 [==============================] - 1s 20us/sample - loss: 0.4945 - acc: 0.8711
Epoch 3/20
60000/60000 [==============================] - 1s 20us/sample - loss: 0.3827 - acc: 0.8950
Epoch 4/20
60000/60000 [==============================] - 1s 20us/sample - loss: 0.3355 - acc: 0.9061
Epoch 5/20
60000/60000 [==============================] - 1s 19us/sample - loss: 0.3066 - acc: 0.9135
Epoch 6/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.2855 - acc: 0.9194
Epoch 7/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.2688 - acc: 0.9236
Epoch 8/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.2544 - acc: 0.9274
Epoch 9/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2420 - acc: 0.9313
Epoch 10/20
60000/60000 [==============================] - 1s 19us/sample - loss: 

In [18]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.16047228739336133
Test Accuracy:0.9513000249862671


#### Compare the results of each model. Which activation function performed best?

The ReLU activation performed the best with a high accuracy score of .95 and a lower test score of .16

#### Model Creation with 3 hidden layers - tanh - 'hinge' loss function

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [19]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='tanh'))
model.add(Dense(64, activation='tanh'))
model.add(Dense(10, activation='softmax'))

In [20]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_10 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_11 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [21]:
model.compile(optimizer='sgd', loss=tf.keras.losses.hinge, metrics=['accuracy'])

In [22]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Epoch 1/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0758 - acc: 0.2210
Epoch 2/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0697 - acc: 0.3514
Epoch 3/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0601 - acc: 0.4108
Epoch 4/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0486 - acc: 0.4717
Epoch 5/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0367 - acc: 0.5127
Epoch 6/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0261 - acc: 0.5458
Epoch 7/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0171 - acc: 0.5823
Epoch 8/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0091 - acc: 0.6177
Epoch 9/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0019 - 

In [23]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.9601568873405456
Test Accuracy:0.8140000104904175


#### Model Creation with 3 hidden layers - Sigmoid - 'hinge' loss function

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [24]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='sigmoid'))
model.add(Dense(64, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

In [25]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_14 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [26]:
model.compile(optimizer='sgd', loss=tf.keras.losses.hinge, metrics=['accuracy'])

In [27]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
60000/60000 [==============================] - 1s 21us/sample - loss: 1.0802 - acc: 0.0992
Epoch 2/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0802 - acc: 0.0992
Epoch 3/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0802 - acc: 0.0992
Epoch 4/20
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0801 - acc: 0.0992
Epoch 5/20
60000/60000 [==============================] - 1s 21us/sample - loss: 1.0801 - acc: 0.0992
Epoch 6/20
60000/60000 [==============================] - 1s 23us/sample - loss: 1.0801 - acc: 0.0992
Epoch 7/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0800 - acc: 0.0992
Epoch 8/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0800 - acc: 0.0992
Epoch 9/20
60000/60000 [==============================] - 1s 23us/sample - loss: 1.0800 - acc: 0.0992
Epoch 10/20
60000/60000 [==============================] - 1s 22us/sample - loss: 

In [28]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:1.0795175422668457
Test Accuracy:0.10090000182390213


#### Model Creation with 3 hidden layers - ReLU - - 'hinge' loss function

Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [29]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [30]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_16 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_17 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [31]:
model.compile(optimizer='sgd', loss=tf.keras.losses.hinge, metrics=['accuracy'])

In [32]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0803 - acc: 0.0692
Epoch 2/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0793 - acc: 0.0981
Epoch 3/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0782 - acc: 0.1246
Epoch 4/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0769 - acc: 0.1414
Epoch 5/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0750 - acc: 0.1476
Epoch 6/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.0723 - acc: 0.1522
Epoch 7/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0690 - acc: 0.1752
Epoch 8/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0658 - acc: 0.2400
Epoch 9/20
60000/60000 [==============================] - 1s 19us/sample - loss: 1.0624 - acc: 0.3135
Epoch 10/20
60000/60000 [==============================] - 1s 21us/sample - loss: 

In [33]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.9845554529190064
Test Accuracy:0.6766999959945679


#### Compare the result of each model with each other. Which activation function did perform better?

Using the hinge loss function, the ReLU activation performed at its best with a accuracy score of .68 and a test score of .98

The hinge loss function however gave lower scores compared to thge previous 3 models.